dlt/sources/helpers/rest_client/paginators.py

```diff
@@ -116,44 +117,55 @@ def __init__(
+              a page contains no result items. Defaults to `True`.
         """
         super().__init__()
-        if total_path is None and maximum_value is None:
-            raise ValueError("Either `total_path` or `maximum_value` must be provided.")
+        if total_path is None and maximum_value is None and not stop_after_empty_page:
+            raise ValueError(
+                "Either `total_path` or `maximum_value` or stop_after_empty_page must be provided."
+            )
         self.param_name = param_name
         self.current_value = initial_value
         self.value_step = value_step
         self.base_index = base_index
         self.maximum_value = maximum_value
         self.total_path = jsonpath.compile_path(total_path) if total_path else None
         self.error_message_items = error_message_items
+        self.stop_after_empty_page = stop_after_empty_page
 
     def init_request(self, request: Request) -> None:
         if request.params is None:
             request.params = {}
 
         request.params[self.param_name] = self.current_value
 
-    def update_state(self, response: Response) -> None:
-        total = None
-        if self.total_path:
-            response_json = response.json()
-            values = jsonpath.find_values(self.total_path, response_json)
-            total = values[0] if values else None
-            if total is None:
-                self._handle_missing_total(response_json)
-
-            try:
-                total = int(total)
-            except ValueError:
-                self._handle_invalid_total(total)
-
-        self.current_value += self.value_step
-
-        if (total is not None and self.current_value >= total + self.base_index) or (
-            self.maximum_value is not None and self.current_value >= self.maximum_value
-        ):
+    def update_state(self, response: Response, data: Optional[List[Any]] = None) -> None:
+        if self._stop_after_this_page(data):
             self._has_next_page = False
+        else:
+            total = None
+            if self.total_path:
+                response_json = response.json()
+                values = jsonpath.find_values(self.total_path, response_json)
+                total = values[0] if values else None
+                if total is None:
+                    self._handle_missing_total(response_json)
+
+                try:
+                    total = int(total)
+                except ValueError:
+                    self._handle_invalid_total(total)
+
+            self.current_value += self.value_step
+
+            if (total is not None and self.current_value >= total + self.base_index) or (
+                self.maximum_value is not None and self.current_value >= self.maximum_value
+            ):
+                self._has_next_page = False
+
+    def _stop_after_this_page(self, data: Optional[List[Any]]) -> bool:
+        return self.stop_after_empty_page and data == []
```

Comment:
I'm wondering if `data == []` is too restrictive. Maybe `not data`? Also the `data` is defined as an `Optional` in `_stop_after_this_page()`, should it have `None` as a default value?

Reply:
Good point! The argument `data` comes from `RESTClient.extract_response()` which always returns a `List[Any]` because its final line is `return cast(List[Any], data)`. 

Thus, I think that this is always a list with 0 or more elements.

But, I prefer providing `data=None` to make it consistent with the other functions, make it more versatile, and explicit.

Reply:
Regarding `data == []` vs `data is None`

The existing test suite calls `update_state()` without the `data` argument and thus `data is None` and the tests break which assert that calling `update_state(a_response)` leaves it in a state which has a next page.  I wonder if there are some custom paginator usages that call `update_state()`. In the dlt code, it is only `RESTClient.paginate()`.

I see the following options:

1. `[]` means that there's no data, `None` means that it should not stop after an empty page even if `stop_after_empty_page is True`. Does not break existing test suite and maybe usage.
2. Both `[]` and `None` mean that there's no data. Breaks existing test suite and maybe usage. Necessary adjustment: 
```py
paginator.update_state(response)
```
becomes 
```py
paginator.update_state(response, data=[{}])
```

I favor option 2 because I find it most clear. But, this requires that we not that in the v1.0 release notes.

What do you think @burnash? Do you see more options?

Reply:
Done in 83bab151